// @rsdoctor/sdk@0.3.2/dist/cjs/sdk/server/apis/graph.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result)
    __defProp(target, key, result);
  return result;
};
var graph_exports = {};
__export(graph_exports, {
  GraphAPI: () => GraphAPI
});
module.exports = __toCommonJS(graph_exports);
var import_types = require("@rsdoctor/types");
var import_base = require("./base");
var import_router = require("../router");
class GraphAPI extends import_base.BaseAPI {
  async getAssetsSummary() {
    const { req } = this.ctx;
    const { withFileContent } = req.body;
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetAssetsSummary, {
      withFileContent
    });
  }
  async getModuleGraphForAsset() {
    const { req } = this.ctx;
    const { assetPath } = req.body;
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetAssetDetails, {
      assetPath
    });
  }
  async getChunksByModuleId() {
    const { req } = this.ctx;
    const { moduleId } = req.body;
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetChunksByModuleId, {
      moduleId
    });
  }
  async getModuleDetails() {
    const { req } = this.ctx;
    const { moduleId } = req.body;
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetModuleDetails, {
      moduleId
    });
  }
  async getDependencyByResolvedRequest() {
    const { req } = this.ctx;
    const { moduleIds = [] } = req.body;
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetModulesByModuleIds, {
      moduleIds
    });
  }
  async getModuleCodeByModuleId() {
    const { req } = this.ctx;
    const { moduleId } = req.body;
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetModuleCodeByModuleId, {
      moduleId
    });
  }
  async getModuleCodeByModuleIds() {
    const { req } = this.ctx;
    const { moduleIds } = req.body;
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetModuleCodeByModuleIds, {
      moduleIds
    });
  }
  async getEntryModulesSummary() {
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetEntryPoints);
  }
}
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetAssetsSummary)
], GraphAPI.prototype, "getAssetsSummary", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetAssetDetails)
], GraphAPI.prototype, "getModuleGraphForAsset", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetChunksByModuleId)
], GraphAPI.prototype, "getChunksByModuleId", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetModuleDetails)
], GraphAPI.prototype, "getModuleDetails", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetModulesByModuleIds)
], GraphAPI.prototype, "getDependencyByResolvedRequest", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetModuleCodeByModuleId)
], GraphAPI.prototype, "getModuleCodeByModuleId", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetModuleCodeByModuleIds)
], GraphAPI.prototype, "getModuleCodeByModuleIds", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetEntryPoints)
], GraphAPI.prototype, "getEntryModulesSummary", 1);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  GraphAPI
});
