// @rsdoctor/sdk@0.3.2/dist/cjs/sdk/server/apis/loader.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result)
    __defProp(target, key, result);
  return result;
};
var loader_exports = {};
__export(loader_exports, {
  LoaderAPI: () => LoaderAPI
});
module.exports = __toCommonJS(loader_exports);
var import_types = require("@rsdoctor/types");
var import_base = require("./base");
var import_router = require("../router");
class LoaderAPI extends import_base.BaseAPI {
  async reportLoader() {
    const { req, sdk } = this.ctx;
    const { body } = req;
    sdk.reportLoader(body);
    return "ok";
  }
  async reportSourceMap() {
    const {
      req: { body },
      sdk
    } = this.ctx;
    sdk.reportSourceMap(body);
    return "ok";
  }
  async getLoaderNames() {
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetLoaderNames);
  }
  async getLoaderChartData() {
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetLoaderChartData);
  }
  async getLoaderFileTree() {
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetLoaderFileTree);
  }
  async getLoaderFileDetails() {
    const { req } = this.ctx;
    const { path } = req.body;
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetLoaderFileDetails, {
      path
    });
  }
  async getLoaderFolderStatistics() {
    const { req } = this.ctx;
    const { folder } = req.body;
    return this.dataLoader.loadAPI(
      import_types.SDK.ServerAPI.API.GetLoaderFolderStatistics,
      { folder }
    );
  }
  async getLoaderFileFirstInput() {
    const { req } = this.ctx;
    const { file } = req.body;
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetLoaderFileFirstInput, {
      file
    });
  }
  async getLoaderFileInputAndOutput() {
    const { req } = this.ctx;
    const { file, loader, loaderIndex } = req.body;
    return this.dataLoader.loadAPI(
      import_types.SDK.ServerAPI.API.GetLoaderFileInputAndOutput,
      { file, loader, loaderIndex }
    );
  }
  async getTileReportHtml() {
    return this.dataLoader.loadAPI(import_types.SDK.ServerAPI.API.GetTileReportHtml, {});
  }
}
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.ReportLoader)
], LoaderAPI.prototype, "reportLoader", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.ReportSourceMap)
], LoaderAPI.prototype, "reportSourceMap", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetLoaderNames)
], LoaderAPI.prototype, "getLoaderNames", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetLoaderChartData)
], LoaderAPI.prototype, "getLoaderChartData", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetLoaderFileTree)
], LoaderAPI.prototype, "getLoaderFileTree", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetLoaderFileDetails)
], LoaderAPI.prototype, "getLoaderFileDetails", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetLoaderFolderStatistics)
], LoaderAPI.prototype, "getLoaderFolderStatistics", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetLoaderFileFirstInput)
], LoaderAPI.prototype, "getLoaderFileFirstInput", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetLoaderFileInputAndOutput)
], LoaderAPI.prototype, "getLoaderFileInputAndOutput", 1);
__decorateClass([
  import_router.Router.post(import_types.SDK.ServerAPI.API.GetTileReportHtml)
], LoaderAPI.prototype, "getTileReportHtml", 1);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LoaderAPI
});
