// @rsdoctor/graph@0.3.2/dist/esm/graph/chunk-graph/chunk.js
class Chunk {
  constructor(id, name, size, initial, entry) {
    this._assets = [];
    this._modules = [];
    this._dependencies = [];
    this._imported = [];
    this.id = id;
    this.name = name;
    this.size = size;
    this.initial = initial;
    this.entry = entry;
  }
  isEntry() {
    return this.entry;
  }
  isChunkEntryModule(module) {
    return module.isEntry || module.getImported().every((item) => !this.hasModule(item));
  }
  hasModule(module) {
    return this._modules.includes(module);
  }
  addModule(module) {
    if (!this.hasModule(module)) {
      this._modules.push(module);
      module.addChunk(this);
    }
  }
  addAsset(asset) {
    this._assets.push(asset);
  }
  addModules(modules) {
    modules.forEach((module) => {
      if (!this.hasModule(module)) {
        this._modules.push(module);
        module.addChunk(this);
      }
    });
  }
  addDependency(dep) {
    if (!this._dependencies.includes(dep)) {
      this._dependencies.push(dep);
    }
  }
  addImported(imported) {
    if (!this._imported.includes(imported)) {
      this._imported.push(imported);
    }
  }
  getAssets() {
    return this._assets.slice();
  }
  getModules() {
    return this._modules.slice();
  }
  getDependencies() {
    return this._dependencies.slice();
  }
  getImported() {
    return this._imported.slice();
  }
  setParsedSize(parsedSize) {
    this._parsedSize = parsedSize;
  }
  toData() {
    return {
      id: this.id,
      name: this.name,
      initial: this.initial,
      size: this.size,
      parsedSize: this._parsedSize || 0,
      entry: this.isEntry(),
      assets: this._assets.map(({ path }) => path),
      modules: this._modules.map(({ id }) => id),
      dependencies: this._dependencies.map(({ id }) => id),
      imported: this._imported.map(({ id }) => id)
    };
  }
}
export {
  Chunk
};
