// @rsdoctor/core@0.3.2/dist/inner-plugins/plugins/base.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var base_exports = {};
__export(base_exports, {
  InternalBasePlugin: () => InternalBasePlugin
});
module.exports = __toCommonJS(base_exports);
var import_constants = require("../constants");
class InternalBasePlugin {
  constructor(scheduler) {
    this.scheduler = scheduler;
  }
  get options() {
    return this.scheduler.options;
  }
  get sdk() {
    return this.scheduler.sdk;
  }
  get tapPostOptions() {
    return (0, import_constants.internalPluginTapPostOptions)(this.name);
  }
  get tapPreOptions() {
    return (0, import_constants.internalPluginTapPreOptions)(this.name);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  InternalBasePlugin
});
