// @rsdoctor/core@0.3.2/dist/build-utils/build/chunks/generateTileGraph.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var generateTileGraph_exports = {};
__export(generateTileGraph_exports, {
  TileGraphReportName: () => TileGraphReportName,
  generateTileGraph: () => generateTileGraph
});
module.exports = __toCommonJS(generateTileGraph_exports);
var import_path = __toESM(require("path"));
var import_logger = require("@rsdoctor/utils/logger");
var import_viewer = require("webpack-bundle-analyzer/lib/viewer");
const TileGraphReportName = "rsdoctor-tile-graph.html";
async function generateJSONReportUtil(bundleStats, opts) {
  await (0, import_viewer.generateReport)(bundleStats, {
    ...opts,
    logger: {
      warn: () => {
      },
      info: () => {
      },
      error: (e) => {
        import_logger.logger.info(`webpack-bundle-analyzer generateReport has error ${e}`);
      }
    }
  });
}
async function generateTileGraph(bundleStats, opts, buildOutputPath) {
  try {
    const { reportFilename } = opts;
    await generateJSONReportUtil(bundleStats, {
      ...opts,
      openBrowser: false,
      bundleDir: buildOutputPath
    });
    return import_path.default.join(buildOutputPath, `${reportFilename}`);
  } catch (e) {
    (0, import_logger.debug)(() => `Generate webpack-bundle-analyzer tile graph has error:${e}`);
    return null;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TileGraphReportName,
  generateTileGraph
});
