// @rsdoctor/utils@0.3.2/dist/cjs/common/rspack.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var rspack_exports = {};
__export(rspack_exports, {
  RspackLoaderInternalPropertyName: () => RspackLoaderInternalPropertyName,
  RspackSummaryCostsDataName: () => RspackSummaryCostsDataName
});
module.exports = __toCommonJS(rspack_exports);
const RspackLoaderInternalPropertyName = "__l__";
var RspackSummaryCostsDataName = /* @__PURE__ */ ((RspackSummaryCostsDataName2) => {
  RspackSummaryCostsDataName2["Bootstrap"] = "bootstrap->rspack:beforeCompile";
  RspackSummaryCostsDataName2["Compile"] = "rspack:beforeCompile->afterCompile";
  RspackSummaryCostsDataName2["Done"] = "rspack:afterCompile->done";
  RspackSummaryCostsDataName2["Minify"] = "rspack:minify(rspack:optimizeChunkAssets)";
  return RspackSummaryCostsDataName2;
})(RspackSummaryCostsDataName || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  RspackLoaderInternalPropertyName,
  RspackSummaryCostsDataName
});
