// @rsdoctor/graph@0.3.2/dist/cjs/graph/module-graph/tree-shaking/module.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var module_exports = {};
__export(module_exports, {
  ModuleGraphModule: () => ModuleGraphModule
});
module.exports = __toCommonJS(module_exports);
var import_export = require("./export");
var import_sideEffect = require("./sideEffect");
var import_variable = require("./variable");
class ModuleGraphModule {
  constructor(module2, graph, dynamic) {
    this.exports = [];
    this.sideEffects = [];
    this.variables = [];
    this.module = module2;
    this._graph = graph;
    if (typeof this._dynamic === "boolean") {
      this._dynamic = dynamic;
    }
  }
  static init() {
    import_export.ExportInfo.init();
    import_sideEffect.SideEffect.init();
    import_variable.Variable.init();
  }
  get dynamic() {
    if (typeof this._dynamic === "boolean") {
      return this._dynamic;
    }
    return this.module.getImported().map((item) => item.getDependencyByModule(this.module)).some((item) => item && item.meta.exportsType === "dynamic");
  }
  addExportInfo(data) {
    this.exports.push(data);
    this._graph.addExportInfo(data);
  }
  addSideEffect(data) {
    this.sideEffects.push(data);
    this._graph.addSideEffect(data);
  }
  addVariable(data) {
    this.variables.push(data);
    this._graph.addVariable(data);
  }
  getExports() {
    return this.exports.slice();
  }
  getSideEffects(name) {
    if (name) {
      return this.sideEffects.filter((item) => item.name === name);
    }
    return this.sideEffects.slice();
  }
  getOwnExports() {
    return this.exports.filter((item) => !item.isReExport);
  }
  getExport(name) {
    return this.exports.find((item) => item.name === name);
  }
  getReExports() {
    return this.exports.filter((item) => item.isReExport);
  }
  getOwnExport(name) {
    return this.getOwnExports().find((item) => item.name === name);
  }
  getReExport(name) {
    return this.getReExports().find((item) => item.name === name);
  }
  toData() {
    return {
      module: this.module.id,
      dynamic: this.dynamic,
      exports: this.exports.map((item) => item.id),
      sideEffects: this.sideEffects.map((item) => item.id),
      variables: this.variables.map((item) => item.id)
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ModuleGraphModule
});
