// @rsdoctor/graph@0.3.2/dist/cjs/graph/module-graph/tree-shaking/sideEffect.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sideEffect_exports = {};
__export(sideEffect_exports, {
  SideEffect: () => SideEffect
});
module.exports = __toCommonJS(sideEffect_exports);
let id = 1;
const _SideEffect = class _SideEffect {
  constructor(name, module2, identifier, fromRequest, originName) {
    this.id = id++;
    this._exports = [];
    this.name = name;
    this.module = module2;
    this.identifier = identifier;
    if (fromRequest) {
      this.fromDependency = this.module.getDependencyByRequest(fromRequest);
    }
    if (originName === _SideEffect.NamespaceSymbol) {
      this.isNameSpace = true;
    } else {
      this.isNameSpace = false;
      this.originName = originName;
    }
  }
  static init() {
    id = 1;
  }
  get variable() {
    if (typeof this._variable !== "undefined") {
      if (this._variable) {
        return this._variable;
      }
      return;
    }
    const result = this.exports[0]?.getRecursiveExport()?.variable;
    this._variable = result ?? false;
    return result;
  }
  get exports() {
    return this._exports?.slice() ?? [];
  }
  setModuleExport(mgm) {
    mgm.getExports().forEach((info) => this.setExportInfo(info));
  }
  setExportInfo(info) {
    this._exports.push(info);
    info.addSideEffect(this);
  }
  toData() {
    const data = {
      id: this.id,
      name: this.name,
      identifier: this.identifier.toData(),
      module: this.module.id
    };
    if (this.fromDependency) {
      data.fromDependency = this.fromDependency.id;
    }
    if (this.isNameSpace) {
      data.isNameSpace = this.isNameSpace;
    }
    if (this.exports.length > 0) {
      data.exports = this.exports.map((item) => item.id);
    }
    if (this.variable) {
      data.variable = this.variable.id;
    }
    return data;
  }
};
/** Name import tags */
_SideEffect.NamespaceSymbol = Symbol("namespace");
let SideEffect = _SideEffect;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SideEffect
});
