// @rsdoctor/core@0.3.2/dist/rules/rules/duplicate-package/types.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  CheckVersion: () => CheckVersion,
  CheckVersionMap: () => CheckVersionMap
});
module.exports = __toCommonJS(types_exports);
var CheckVersion = /* @__PURE__ */ ((CheckVersion2) => {
  CheckVersion2[CheckVersion2["null"] = 0] = "null";
  CheckVersion2[CheckVersion2["prerelease"] = 1] = "prerelease";
  CheckVersion2[CheckVersion2["prepatch"] = 16] = "prepatch";
  CheckVersion2[CheckVersion2["patch"] = 256] = "patch";
  CheckVersion2[CheckVersion2["preminor"] = 4096] = "preminor";
  CheckVersion2[CheckVersion2["minor"] = 65536] = "minor";
  CheckVersion2[CheckVersion2["premajor"] = 1048576] = "premajor";
  CheckVersion2[CheckVersion2["major"] = 16777216] = "major";
  return CheckVersion2;
})(CheckVersion || {});
var CheckVersionMap = /* @__PURE__ */ ((CheckVersionMap2) => {
  CheckVersionMap2[CheckVersionMap2["null"] = 0] = "null";
  CheckVersionMap2[CheckVersionMap2["prerelease"] = 1] = "prerelease";
  CheckVersionMap2[CheckVersionMap2["prepatch"] = 17] = "prepatch";
  CheckVersionMap2[CheckVersionMap2["patch"] = 273] = "patch";
  CheckVersionMap2[CheckVersionMap2["preminor"] = 4369] = "preminor";
  CheckVersionMap2[CheckVersionMap2["minor"] = 69905] = "minor";
  CheckVersionMap2[CheckVersionMap2["premajor"] = 1118481] = "premajor";
  CheckVersionMap2[CheckVersionMap2["major"] = 17895697] = "major";
  return CheckVersionMap2;
})(CheckVersionMap || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CheckVersion,
  CheckVersionMap
});
