// @rsdoctor/utils@0.3.2/dist/cjs/common/graph/assets.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var assets_exports = {};
__export(assets_exports, {
  diffAssetsByExtensions: () => diffAssetsByExtensions,
  diffSize: () => diffSize,
  filterAssets: () => filterAssets,
  filterAssetsByExtensions: () => filterAssetsByExtensions,
  formatAssetName: () => formatAssetName,
  getAssetDetails: () => getAssetDetails,
  getAssetsDiffResult: () => getAssetsDiffResult,
  getAssetsSizeInfo: () => getAssetsSizeInfo,
  getAssetsSummary: () => getAssetsSummary,
  getInitialAssetsSizeInfo: () => getInitialAssetsSizeInfo,
  isAssetMatchExtension: () => isAssetMatchExtension,
  isAssetMatchExtensions: () => isAssetMatchExtensions,
  isInitialAsset: () => isInitialAsset
});
module.exports = __toCommonJS(assets_exports);
var import_types = require("@rsdoctor/types");
var import_chunk = require("./chunk");
var import_modules = require("./modules");
const EXT = "js|css|html";
const hashPattern = /[a-z|A-Z|0-9]{4,32}/;
const hashSeparatorPattern = /[-|.]/;
const fileExtensionPattern = /(?:\.[a-z｜A-Z|0-9]{2,}){1,}/;
const filenamePattern = new RegExp(
  `(.*)${hashSeparatorPattern.source}${hashPattern.source}(${fileExtensionPattern.source})$`
);
function formatAssetName(assetName, fileConfig) {
  const splitFilesList = fileConfig?.split(".");
  let outputFileTailName = "";
  let unHashedFileName = assetName;
  if (splitFilesList?.length && splitFilesList.length >= 3 && splitFilesList[splitFilesList.length - 2]?.indexOf("[") < 0 && EXT.indexOf(splitFilesList[splitFilesList.length - 1]) > -1) {
    outputFileTailName = splitFilesList[splitFilesList.length - 2];
    const _regPattern = /(.*)(\.[a-f0-9]{4,32})([^.]*.[^.]+){2,}/g;
    unHashedFileName = assetName.replace(_regPattern, "$1");
    return `${unHashedFileName}.${outputFileTailName}.${assetName.substring(
      assetName.lastIndexOf(".") + 1
    )}`;
  }
  return assetName.replace(filenamePattern, "$1$2");
}
function isAssetMatchExtension(asset, ext) {
  return asset.path.slice(-ext.length) === ext;
}
function isAssetMatchExtensions(asset, exts) {
  if (!exts.length)
    return false;
  return exts.some((ext) => isAssetMatchExtension(asset, ext));
}
function filterAssetsByExtensions(assets, exts) {
  if (typeof exts === "string")
    return assets.filter((e) => isAssetMatchExtension(e, exts));
  if (Array.isArray(exts)) {
    return assets.filter((e) => isAssetMatchExtensions(e, exts));
  }
  return [];
}
function filterAssets(assets, filterOrExtensions) {
  if (filterOrExtensions) {
    if (typeof filterOrExtensions === "function") {
      assets = assets.filter(filterOrExtensions);
    } else {
      assets = filterAssetsByExtensions(assets, filterOrExtensions);
    }
  }
  return assets;
}
function getAssetsSizeInfo(assets, chunks, options = {}) {
  const { withFileContent = true, filterOrExtensions } = options;
  assets = assets.filter(
    (e) => !isAssetMatchExtensions(e, import_types.Constants.MapExtensions)
  );
  if (filterOrExtensions) {
    assets = filterAssets(assets, filterOrExtensions);
  }
  if (assets.length) {
    return {
      count: assets.length,
      size: assets.reduce((t, c) => t + c.size, 0),
      files: assets.map((e) => ({
        path: e.path,
        size: e.size,
        initial: isInitialAsset(e, chunks),
        content: withFileContent ? e.content : void 0
      }))
    };
  }
  return {
    count: 0,
    size: 0,
    files: []
  };
}
function isInitialAsset(asset, chunks) {
  const cks = (0, import_chunk.getChunksByAsset)(asset, chunks);
  return cks.some((e) => Boolean(e.initial));
}
function getInitialAssetsSizeInfo(assets, chunks, options = {}) {
  if (options.filterOrExtensions) {
    assets = filterAssets(assets, options.filterOrExtensions);
  }
  return getAssetsSizeInfo(assets, chunks, {
    ...options,
    filterOrExtensions: (asset) => isInitialAsset(asset, chunks)
  });
}
function getAssetsDiffResult(baseline, current) {
  return {
    all: {
      total: diffAssetsByExtensions(baseline, current)
    },
    js: {
      total: diffAssetsByExtensions(baseline, current, import_types.Constants.JSExtension),
      initial: diffAssetsByExtensions(
        baseline,
        current,
        import_types.Constants.JSExtension,
        true
      )
    },
    css: {
      total: diffAssetsByExtensions(baseline, current, import_types.Constants.CSSExtension),
      initial: diffAssetsByExtensions(
        baseline,
        current,
        import_types.Constants.CSSExtension,
        true
      )
    },
    imgs: {
      total: diffAssetsByExtensions(baseline, current, import_types.Constants.ImgExtensions)
    },
    html: {
      total: diffAssetsByExtensions(baseline, current, import_types.Constants.HtmlExtension)
    },
    media: {
      total: diffAssetsByExtensions(
        baseline,
        current,
        import_types.Constants.MediaExtensions
      )
    },
    fonts: {
      total: diffAssetsByExtensions(
        baseline,
        current,
        import_types.Constants.FontExtensions
      )
    },
    others: {
      total: diffAssetsByExtensions(
        baseline,
        current,
        (asset) => !isAssetMatchExtensions(
          asset,
          [
            import_types.Constants.JSExtension,
            import_types.Constants.CSSExtension,
            import_types.Constants.HtmlExtension
          ].concat(
            import_types.Constants.ImgExtensions,
            import_types.Constants.MediaExtensions,
            import_types.Constants.FontExtensions,
            import_types.Constants.MapExtensions
          )
        )
      )
    }
  };
}
function diffSize(bSize, cSize) {
  const isEqual = bSize === cSize;
  const percent = isEqual ? 0 : bSize === 0 ? 100 : Math.abs(cSize - bSize) / bSize * 100;
  const state = isEqual ? import_types.Client.RsdoctorClientDiffState.Equal : bSize > cSize ? import_types.Client.RsdoctorClientDiffState.Down : import_types.Client.RsdoctorClientDiffState.Up;
  return { percent, state };
}
function diffAssetsByExtensions(baseline, current, filterOrExtensions, isInitial = false) {
  const { size: bSize, count: bCount } = isInitial ? getInitialAssetsSizeInfo(baseline.assets, baseline.chunks, {
    filterOrExtensions
  }) : getAssetsSizeInfo(baseline.assets, baseline.chunks, {
    filterOrExtensions
  });
  let cSize;
  let cCount;
  if (baseline === current) {
    cSize = bSize;
    cCount = bCount;
  } else {
    const { size, count } = isInitial ? getInitialAssetsSizeInfo(current.assets, current.chunks, {
      filterOrExtensions
    }) : getAssetsSizeInfo(current.assets, current.chunks, {
      filterOrExtensions
    });
    cSize = size;
    cCount = count;
  }
  const { percent, state } = diffSize(bSize, cSize);
  return {
    size: {
      baseline: bSize,
      current: cSize
    },
    count: {
      baseline: bCount,
      current: cCount
    },
    percent,
    state
  };
}
function getAssetsSummary(assets, chunks, options = {}) {
  const jsOpt = {
    ...options,
    filterOrExtensions: import_types.Constants.JSExtension
  };
  const cssOpt = {
    ...options,
    filterOrExtensions: import_types.Constants.CSSExtension
  };
  const imgOpt = {
    ...options,
    filterOrExtensions: import_types.Constants.ImgExtensions
  };
  const htmlOpt = {
    ...options,
    filterOrExtensions: import_types.Constants.HtmlExtension
  };
  const mediaOpt = {
    ...options,
    filterOrExtensions: import_types.Constants.MediaExtensions
  };
  const fontOpt = {
    ...options,
    filterOrExtensions: import_types.Constants.FontExtensions
  };
  const otherOpt = {
    ...options,
    filterOrExtensions: (asset) => !isAssetMatchExtensions(
      asset,
      [
        import_types.Constants.JSExtension,
        import_types.Constants.CSSExtension,
        import_types.Constants.HtmlExtension
      ].concat(
        import_types.Constants.ImgExtensions,
        import_types.Constants.MediaExtensions,
        import_types.Constants.FontExtensions,
        import_types.Constants.MapExtensions
      )
    )
  };
  return {
    all: {
      total: getAssetsSizeInfo(assets, chunks, options)
    },
    js: {
      total: getAssetsSizeInfo(assets, chunks, jsOpt),
      initial: getInitialAssetsSizeInfo(assets, chunks, jsOpt)
    },
    css: {
      total: getAssetsSizeInfo(assets, chunks, cssOpt),
      initial: getInitialAssetsSizeInfo(assets, chunks, cssOpt)
    },
    imgs: {
      total: getAssetsSizeInfo(assets, chunks, imgOpt)
    },
    html: {
      total: getAssetsSizeInfo(assets, chunks, htmlOpt)
    },
    media: {
      total: getAssetsSizeInfo(assets, chunks, mediaOpt)
    },
    fonts: {
      total: getAssetsSizeInfo(assets, chunks, fontOpt)
    },
    others: {
      total: getAssetsSizeInfo(assets, chunks, otherOpt)
    }
  };
}
function getAssetDetails(assetPath, assets, chunks, modules) {
  const asset = assets.find((e) => e.path === assetPath);
  return {
    asset,
    chunks: (0, import_chunk.getChunksByAsset)(asset, chunks),
    modules: (0, import_modules.getModulesByAsset)(asset, chunks, modules)
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  diffAssetsByExtensions,
  diffSize,
  filterAssets,
  filterAssetsByExtensions,
  formatAssetName,
  getAssetDetails,
  getAssetsDiffResult,
  getAssetsSizeInfo,
  getAssetsSummary,
  getInitialAssetsSizeInfo,
  isAssetMatchExtension,
  isAssetMatchExtensions,
  isInitialAsset
});
