// @rsdoctor/types@0.3.2/dist/sdk/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToDataType = exports.ModuleKind = exports.DependencyKind = void 0;
var DependencyKind;
(function (DependencyKind) {
    DependencyKind[DependencyKind["Unknown"] = 0] = "Unknown";
    DependencyKind[DependencyKind["ImportStatement"] = 1] = "ImportStatement";
    DependencyKind[DependencyKind["DynamicImport"] = 2] = "DynamicImport";
    DependencyKind[DependencyKind["RequireCall"] = 3] = "RequireCall";
    DependencyKind[DependencyKind["AMDRequire"] = 4] = "AMDRequire";
})(DependencyKind || (exports.DependencyKind = DependencyKind = {}));
/** Module type */
var ModuleKind;
(function (ModuleKind) {
    /** Normal module */
    ModuleKind[ModuleKind["Normal"] = 0] = "Normal";
    /** Aggregation module */
    ModuleKind[ModuleKind["Concatenation"] = 1] = "Concatenation";
})(ModuleKind || (exports.ModuleKind = ModuleKind = {}));
var ToDataType;
(function (ToDataType) {
    /** Normal mode */
    ToDataType[ToDataType["Normal"] = 0] = "Normal";
    /** Lite mode */
    ToDataType[ToDataType["Lite"] = 1] = "Lite";
    /** Lite & No Assets Code mode */
    ToDataType[ToDataType["LiteAndNoAsset"] = 2] = "LiteAndNoAsset";
    /** All Data, include module transformed data. */
    ToDataType[ToDataType["All"] = 3] = "All";
})(ToDataType || (exports.ToDataType = ToDataType = {}));
