// @rsdoctor/graph@0.3.2/dist/cjs/graph/module-graph/module.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var module_exports = {};
__export(module_exports, {
  Module: () => Module
});
module.exports = __toCommonJS(module_exports);
var import_types = require("@rsdoctor/types");
var import_path = __toESM(require("path"));
var import_lodash = require("lodash");
var import_dependency = require("./dependency");
var import_statement = require("./statement");
var import_utils = require("./utils");
let id = 1;
class Module {
  constructor(webpackId, path2, isEntry = false, kind = import_types.SDK.ModuleKind.Normal) {
    this.source = {
      source: "",
      transformed: "",
      parsedSource: ""
    };
    this.size = {
      sourceSize: 0,
      transformedSize: 0,
      parsedSize: 0
    };
    this.chunks = [];
    this.dependencies = [];
    this.imported = [];
    this.modules = [];
    this.concatenationModules = [];
    this.meta = {
      hasSetEsModuleStatement: false,
      strictHarmonyModule: false
    };
    this.id = id++;
    this.webpackId = webpackId;
    this.path = path2;
    this.isEntry = isEntry;
    this.kind = kind;
  }
  static init() {
    id = 1;
  }
  get rootModule() {
    return this.modules.find((item) => item.path === this.path);
  }
  get isPreferSource() {
    if (typeof this._isPreferSource === "boolean") {
      return this._isPreferSource;
    }
    const result = this.source.source.length > 0 && this.source.source !== "test code" && Boolean(this.sourceMap);
    this._isPreferSource = result;
    return result;
  }
  getChunks() {
    return this.chunks.slice();
  }
  addChunk(chunk) {
    if (!this.chunks.includes(chunk)) {
      this.chunks.push(chunk);
      chunk.addModule(this);
    }
  }
  removeChunk(chunk) {
    this.chunks = this.chunks.filter((item) => item !== chunk);
  }
  getDependencies() {
    return this.dependencies.slice();
  }
  getDependencyByRequest(request) {
    return this.dependencies.find((item) => item.request === request);
  }
  getDependencyByModule(module2) {
    return this.dependencies.find(
      (item) => item.originDependency === module2 || item.dependency === module2
    );
  }
  addDependency(request, module2, kind, statements) {
    const dep = new import_dependency.Dependency(request, this, module2, kind, statements);
    if (this.dependencies.every((item) => !item.isSameWithoutStatements(dep))) {
      this.dependencies.push(dep);
      module2.addImported(this);
      if (module2.rootModule) {
        module2.rootModule.addImported(this);
      }
      return dep;
    }
  }
  removeDependency(dep) {
    this.dependencies = this.dependencies.filter((item) => item === dep);
  }
  removeDependencyByModule(module2) {
    const dep = this.getDependencyByModule(module2);
    if (dep) {
      this.removeDependency(dep);
    }
  }
  getImported() {
    return this.imported.slice();
  }
  addImported(module2) {
    if (!this.imported.includes(module2)) {
      this.imported.push(module2);
    }
  }
  removeImported(module2) {
    this.imported = this.imported.filter((item) => item === module2);
  }
  setProgram(program) {
    this.program = program;
  }
  getProgram() {
    return this.program;
  }
  setSource(input) {
    const { source } = this;
    source.source = input.source ?? source.source;
    source.transformed = input.transformed ?? source.transformed;
    source.parsedSource = input.parsedSource ?? source.parsedSource;
  }
  getSource(type = import_types.SDK.ToDataType.Normal) {
    if (type === import_types.SDK.ToDataType.Lite || type === import_types.SDK.ToDataType.LiteAndNoAsset) {
      return {
        source: "",
        transformed: "",
        parsedSource: this.isPreferSource ? "" : this.source.parsedSource
      };
    }
    if (type === import_types.SDK.ToDataType.All) {
      return {
        source: this.source.source,
        transformed: this.source.transformed,
        parsedSource: this.isPreferSource ? "" : this.source.parsedSource
      };
    }
    return {
      source: this.source.source,
      transformed: "",
      parsedSource: this.isPreferSource ? "" : this.source.parsedSource
    };
  }
  setSourceMap(sourceMap) {
    this.sourceMap = sourceMap;
  }
  getSourceMap() {
    return this.sourceMap;
  }
  setSize(input) {
    const { size } = this;
    size.sourceSize = input.sourceSize ?? size.sourceSize;
    size.transformedSize = input.transformedSize ?? size.transformedSize;
    size.parsedSize = input.parsedSize ?? size.parsedSize;
  }
  getSize() {
    return { ...this.size };
  }
  getStatement(transformed) {
    return new import_statement.Statement(this, {
      source: this.getSourceRange(transformed),
      transformed: {
        start: { ...transformed.start },
        end: transformed.end ? { ...transformed.end } : void 0
      }
    });
  }
  getSourceRange(transformed) {
    const { sourceMap } = this;
    if (!sourceMap) {
      return;
    }
    const source = {
      start: {}
    };
    const startInSource = sourceMap.originalPositionFor({
      line: transformed.start.line ?? 0,
      column: transformed.start.column ?? 0,
      // The largest lower bound.
      bias: 1
    });
    if ((0, import_lodash.isNumber)(startInSource.line)) {
      source.start = {
        line: startInSource.line,
        column: startInSource.column ?? void 0
      };
    }
    if (transformed.end) {
      const endInSource = sourceMap.originalPositionFor({
        line: transformed.end.line ?? 0,
        column: transformed.end.column ?? 0
        // The smallest lower bound
        // bias: 2,
      });
      if ((0, import_lodash.isNumber)(endInSource.line)) {
        source.end = {
          line: endInSource.line,
          column: endInSource.column ?? void 0
        };
      }
    }
    return source;
  }
  addNormalModule(module2) {
    if (!this.modules.includes(module2)) {
      this.modules.push(module2);
      module2.addConcatenationModule(this);
    }
  }
  getNormalModules() {
    return this.modules.slice();
  }
  addConcatenationModule(module2) {
    if (!this.concatenationModules.includes(module2)) {
      this.concatenationModules.push(module2);
    }
  }
  getConcatenationModules() {
    return this.concatenationModules.slice();
  }
  toData(contextPath) {
    const { isPreferSource } = this;
    const moduleName = (0, import_utils.getModuleName)(this.webpackId);
    const data = {
      id: this.id,
      webpackId: contextPath && moduleName.indexOf(".") > 0 ? import_path.default.relative(contextPath, moduleName) : this.webpackId,
      path: this.path,
      isPreferSource,
      dependencies: this.dependencies.map((item) => item.id),
      imported: this.imported.map((item) => item.id),
      chunks: this.chunks.map((item) => item.id),
      size: this.getSize(),
      kind: this.kind
    };
    if (this.meta.hasSetEsModuleStatement || this.meta.strictHarmonyModule) {
      data.meta = {};
      if (this.meta.hasSetEsModuleStatement) {
        data.meta.hasSetEsModuleStatement = true;
      }
      if (this.meta.strictHarmonyModule) {
        data.meta.strictHarmonyModule = true;
      }
    }
    if (this.isEntry) {
      data.isEntry = this.isEntry;
    }
    if (this.modules.length > 0) {
      data.modules = this.modules.map((item) => item.id);
    }
    if (this.rootModule) {
      data.rootModule = this.rootModule.id;
    }
    if (this.concatenationModules.length > 0) {
      data.concatenationModules = this.concatenationModules.map(
        (data2) => data2.id
      );
    }
    return data;
  }
}
Module.kind = import_types.SDK.ModuleKind;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Module
});
