// @rsdoctor/core@0.3.2/dist/build-utils/common/module-graph/compatible.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var compatible_exports = {};
__export(compatible_exports, {
  getPositionByStatsLocation: () => getPositionByStatsLocation,
  isWebpack5orRspack: () => isWebpack5orRspack
});
module.exports = __toCommonJS(compatible_exports);
function getPositionByStatsLocation(loc) {
  function formatPosition(loc2) {
    const [line, column] = loc2.split(":");
    return { line: Number(line), column: Number(column) };
  }
  if (loc.includes("-")) {
    const positionString = loc.split("-");
    const start = formatPosition(positionString[0]);
    const end = formatPosition(positionString[1]);
    return {
      start,
      end: Number.isNaN(end.column) ? {
        line: start.line,
        column: end.line
      } : end
    };
  }
  if (loc.includes(":")) {
    return {
      start: formatPosition(loc)
    };
  }
  if (loc.includes("[")) {
    const result = loc.match(/\[(\d+)\]/);
    return {
      start: {
        index: Number(result[1])
      }
    };
  }
}
function isWebpack5orRspack(compilation) {
  return "moduleGraph" in compilation && Boolean(compilation.moduleGraph);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getPositionByStatsLocation,
  isWebpack5orRspack
});
