// @rsdoctor/types@0.3.2/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RsdoctorClientUrl = exports.RsdoctorProcessEnvDebugKey = exports.RsdoctorMonitorDocBId = exports.RsdoctorMonitorWebBId = exports.RsdoctorMonitorNodeBId = exports.StatsFilePath = exports.RsdoctorOutputManifestPath = exports.RsdoctorOutputManifest = exports.RsdoctorOutputFolder = exports.MapExtensions = exports.FontExtensions = exports.MediaExtensions = exports.ImgExtensions = exports.HtmlExtension = exports.CSSExtension = exports.JSExtension = void 0;
exports.JSExtension = '.js';
exports.CSSExtension = '.css';
exports.HtmlExtension = '.html';
exports.ImgExtensions = [
    '.png',
    '.jpg',
    '.svg',
    '.webp',
    '.jpeg',
    '.gif',
    '.bmp',
];
exports.MediaExtensions = [
    '.mp3',
    '.mp4',
    '.avi',
    '.wav',
    '.flv',
    '.mov',
    '.mpg',
    '.mpeg',
];
exports.FontExtensions = ['.ttf', '.fnt', '.fon', '.otf', '.woff', '.woff2'];
exports.MapExtensions = ['.js.map', '.ts.map', '.LICENSE.txt'];
exports.RsdoctorOutputFolder = '.rsdoctor';
exports.RsdoctorOutputManifest = 'manifest.json';
exports.RsdoctorOutputManifestPath = `${exports.RsdoctorOutputFolder}/${exports.RsdoctorOutputManifest}`;
exports.StatsFilePath = 'dist/stats.json';
exports.RsdoctorMonitorNodeBId = 'Rsdoctor';
exports.RsdoctorMonitorWebBId = 'Rsdoctor-Client';
exports.RsdoctorMonitorDocBId = 'Rsdoctor-Doc';
exports.RsdoctorProcessEnvDebugKey = 'RSDOCTOR_DEBUG';
exports.RsdoctorClientUrl = ''; // TODO: rsdoctor client url
