// @rsdoctor/types@0.3.2/dist/manifest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RsdoctorManifestClientConstant = exports.RsdoctorManifestClientRoutes = void 0;
var RsdoctorManifestClientRoutes;
(function (RsdoctorManifestClientRoutes) {
    RsdoctorManifestClientRoutes["Overall"] = "Overall";
    RsdoctorManifestClientRoutes["WebpackLoaders"] = "Compile.WebpackLoaders";
    RsdoctorManifestClientRoutes["ModuleResolve"] = "Compile.ModuleResolve";
    RsdoctorManifestClientRoutes["WebpackPlugins"] = "Compile.WebpackPlugins";
    RsdoctorManifestClientRoutes["BundleSize"] = "Bundle.BundleSize";
    RsdoctorManifestClientRoutes["ModuleGraph"] = "Bundle.ModuleGraph";
    RsdoctorManifestClientRoutes["TreeShaking"] = "Bundle.TreeShaking";
})(RsdoctorManifestClientRoutes || (exports.RsdoctorManifestClientRoutes = RsdoctorManifestClientRoutes = {}));
var RsdoctorManifestClientConstant;
(function (RsdoctorManifestClientConstant) {
    RsdoctorManifestClientConstant["WindowPropertyForManifestUrl"] = "__DEVTOOLS_MANIFEST_URL__";
})(RsdoctorManifestClientConstant || (exports.RsdoctorManifestClientConstant = RsdoctorManifestClientConstant = {}));
