// @rsdoctor/graph@0.3.2/dist/cjs/graph/chunk-graph/asset.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var asset_exports = {};
__export(asset_exports, {
  Asset: () => Asset
});
module.exports = __toCommonJS(asset_exports);
var import_types = require("@rsdoctor/types");
class Asset {
  constructor(path, size, chunks, content) {
    this.path = path;
    this.size = size;
    this.chunks = chunks;
    this.content = content;
  }
  toData(types) {
    return {
      path: this.path,
      size: this.size,
      chunks: this.chunks?.map((ck) => ck.id),
      content: types === import_types.SDK.ToDataType.LiteAndNoAsset ? "" : this.content
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Asset
});
